---
title: "p0094 report"
author: "Jinlong Ru"
date: "Compiled at `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'CET')` CET"
output: word_document
bibliography: refs.bib
params:
  name: "99-report" # change if you rename file
---

```{r here, message=FALSE, include=FALSE}
here::i_am(paste0(params$name, ".Rmd"), uuid = "3b73f5e3-f078-42f1-96f5-ffeb3fb1b912")
# create or *empty* the target directory, used to write this file's data: 
projthis::proj_create_dir_target(params$name, clean = TRUE)

# function to get path to target directory: path_target("sample.csv")
path_target <- projthis::proj_path_target(params$name)

# function to get path to previous data: path_source("00-import", "sample.csv")
path_source <- projthis::proj_path_source(params$name)
```

Raw sequencing reads were ﬁltered using fastp [@2018-Fastp_Chen] to remove adaptors and low-quality bases. Remaining reads were deduplicated using dedupe.sh from the bbmap suite version 38.79 [@2014-BBMap_Bushnell]. The obtained clean reads were assembled using metaSPAdes with k-mer size set to 21,33,55,77,99 [@2017-MetaSPAdes_Nurk]. Contigs longer than 3000 bp from all samples were pooled into a contig library (cclib). CheckV (v0.8.1) [@2020-CheckV_Nayfach] was used to removing flanking host region of proviruses, and evaluate completeness and quality of putative viral contigs. To further reduce the redundancy of the contig library, we clustered contigs into a non-redundant contig library (nrclib) following the “rapid genome clustering based on pairwise ANI” protocol in CheckV [@2020-CheckV_Nayfach]. The longest contig from each cluster sharing more than 95% identity and 80% coverage were selected as the representative in the nrclib. Clean reads were mapped to nrclib using minimap2 [@2018-Minimap2_Li], and abundance of each contig were calculated using CoverM (v0.6.1) (https://github.com/wwood/CoverM). VirSorter2 (v2.2.3) [@2021-VirSorter2_Guo] was used to identify viral contigs in nrclib. Contigs classified as categories 1 and 2 by VirSorter2 were selected for further analyses. DRAM-v [@2020-DRAM_Shaffer] was used to annotate viral contigs. Taxonomy annotation of viral contigs were annotated using two approaches. First, genus-level taxonomy information were obtained using vConTACT2 [@2019-Taxonomic_BinJang]. Since most of viral contigs cannot be annotated using vConTACT2, in the second step, we searched NCBI viral RefSeq database using MMSeqs2 taxonomy module [@2021-Fast_Mirdita]. Using this strategy, we assigned taxonomy to 2133/2455 viral contigs. Hosts of viruses were predicted using IPHoP [@2022-IPHoP_Roux] with default parameters. Lifestyle of viruses were predicted using BACPHLIP v0.9.6 [@2021-BACPHLIP_Hockenberry].

### References
